### PR TITLE
fix: log warn when external url is not fully qualified (#4098)

### LINKED ISSUE #4098: Improve error message when excluding https:// from the URL in configmap
# Summary
URL validation for argocd-server configmap and print a error message that the URL does not contain any protocol prefix.

# Motivation

When configuring Dex config for SSO in the configmap argocd-cm is it required to provide a URL to the ArgoCD instance. 

If this URL does not contain explicitly the HTTPS protocol every SSO login attempt will generate a `Failed to query provider "argocd-server.io/api/dex": 400 Bad Request: 400 Bad Request` message and in the argocd-server log `time="2020-08-14T12:19:03Z" level=info msg="Initializing OIDC provider (issuer: argocd-server.io/api/dex)"`. 

These two log entries are not good enough for debugging what is wrong when in fact it is required that the URL must be formatted like this:

Example 

BAD configmap 
```
---
kind: ConfigMap
apiVersion: v1
metadata:
  labels:
    app.kubernetes.io/name: argocd-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-cm
  namespace: argocd
data:
  url: argocd-server.io
  dex.config: |
    logger:
      level: debug
      format: json
    connectors:
    - type: saml
      name: SAML
      id: saml
      config:
        ssoURL: https://identity-provider/idp/profile/SAML2/POST/SSO
        entityIssuer: https://argocd-server.io/api/dex
        ca: /opt/cert/ca.pem
        usernameAttr: name
        emailAttr: email
        groupsAttr: groups
```

GOOD configmap 
```
---
kind: ConfigMap
apiVersion: v1
metadata:
  labels:
    app.kubernetes.io/name: argocd-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-cm
  namespace: argocd
data:
  url: https://argocd-server.io
  dex.config: |
    logger:
      level: debug
      format: json
    connectors:
    - type: saml
      name: SAML
      id: saml
      config:
        ssoURL: https://identity-provider/idp/profile/SAML2/POST/SSO
        entityIssuer: https://argocd-server.io/api/dex
        ca: /opt/cert/ca.pem
        usernameAttr: name
        emailAttr: email
        groupsAttr: groups
```

# Proposal

When argocd-server loads the configmap `argocd-cm` do a validation of the URL (data.url) provided and check if it has a protocol prefixed or not and log accordingly.

### PR DESCRIPTION
**Pull request by olegsu**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] ~~I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.~~
* [ ] ~~Optional. My organization is added to USERS.md.~~
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 

Fixes #4098

### CODE REVIEW

File: util/settings/settings.go [hunk 1]
```diff
@@ -790,14 +790,33 @@ func (mgr *SettingsManager) ensureSynced(forceResync bool) error {
...
 	settings.StatusBadgeEnabled = argoCDCM.Data[statusBadgeEnabledKey] == "true"
 	settings.AnonymousUserEnabled = argoCDCM.Data[anonymousUserEnabledKey] == "true"
 	settings.UiCssURL = argoCDCM.Data[settingUiCssURLKey]
+	if err := validateExternalURL(argoCDCM.Data[settingURLKey]); err != nil {
+		log.Warnf("Failed to validate URL in configmap: %v", err)
+	}
+	settings.URL = argoCDCM.Data[settingURLKey]
+}
+
+// validateExternalURL ensures the extenral URL that is set on the configmap is valid
```
Comment (mekstrem) [inline util/settings/settings.go:807]: Small typo: s/extenral/external/

Comment (olegsu) [reply util/settings/settings.go]: Thanks!

**Comment by olegsu** [comment]
image

File: util/settings/settings.go [hunk 2]
```diff
@@ -790,14 +790,33 @@ func (mgr *SettingsManager) ensureSynced(forceResync bool) error {
...
+// validateExternalURL ensures the external URL that is set on the configmap is valid
+func validateExternalURL(u string) error {
+	if u == "" {
+		return nil
+	}
+	URL, err := url.Parse(u)
+	if err != nil {
+		return fmt.Errorf("Failed to parse URL: %v", err)
+	}
+	if URL.Scheme == "" {
```
Comment (jannfis) [inline util/settings/settings.go:816]: I think we can explicitly test for `URL.Scheme != "http" && URL.Scheme != "https"` here, as we also mention both in the warnings, so user can't sneak in a `gopher://` URL :-)

**Review by jannfis — COMMENTED** [review]
LGTM with a minor comment - what do you think?

File: util/settings/settings.go (hunk 2 above)
Comment (olegsu) [reply util/settings/settings.go]: Sounds good!

**Review by jannfis — APPROVED** [review]
LGTM, thank you for the PR!

**Comment by jannfis** [comment]
@alexmt Maybe this one can make it into one of the 1.7.x releases.

**Pull request merged**